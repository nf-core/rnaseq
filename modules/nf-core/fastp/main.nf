process FASTP {
    tag "$meta.id"
    label 'process_medium'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/fastp:0.23.4--h5f740d0_0' :
        'biocontainers/fastp:0.23.4--h5f740d0_0' }"

    input:
    tuple val(meta), path(reads)
    path  adapter_fasta
    val   save_trimmed_fail
    val   save_merged

    output:
    tuple val(meta), path('*.fastp.fastq.gz') , optional:true, emit: reads
    tuple val(meta), path('*.json')           , emit: json
    tuple val(meta), path('*.html')           , emit: html
    tuple val(meta), path('*.log')            , emit: log
    path "versions.yml"                       , emit: versions
    tuple val(meta), path('*.fail.fastq.gz')  , optional:true, emit: reads_fail
    tuple val(meta), path('*.merged.fastq.gz'), optional:true, emit: reads_merged

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def adapter_list = adapter_fasta ? "--adapter_fasta ${adapter_fasta}" : ""
    def fail_fastq = save_trimmed_fail && meta.single_end ? "--failed_out ${prefix}.fail.fastq.gz" : save_trimmed_fail && !meta.single_end ? "--unpaired1 ${prefix}_1.fail.fastq.gz --unpaired2 ${prefix}_2.fail.fastq.gz" : ''
    // Added soft-links to original fastqs for consistent naming in MultiQC
    // Use single ended for interleaved. Add --interleaved_in in config.
    if ( task.ext.args?.contains('--interleaved_in') ) {
        """
        [ ! -f  ${prefix}.fastq.gz ] && ln -sf $reads ${prefix}.fastq.gz

        fastp \\
            --stdout \\
            --in1 ${prefix}.fastq.gz \\
            --thread $task.cpus \\
            --json ${prefix}.fastp.json \\
            --html ${prefix}.fastp.html \\
            $adapter_list \\
            $fail_fastq \\
            $args \\
            2> >(tee ${prefix}.fastp.log >&2) \\
        | gzip -c > ${prefix}.fastp.fastq.gz

        cat <<-END_VERSIONS > versions.yml
        "${task.process}":
            fastp: \$(fastp --version 2>&1 | sed -e "s/fastp //g")
        END_VERSIONS
        """
    } else if (meta.single_end) {
        """
        [ ! -f  ${prefix}.fastq.gz ] && ln -sf $reads ${prefix}.fastq.gz

        fastp \\
            --in1 ${prefix}.fastq.gz \\
            --out1  ${prefix}.fastp.fastq.gz \\
            --thread $task.cpus \\
            --json ${prefix}.fastp.json \\
            --html ${prefix}.fastp.html \\
            $adapter_list \\
            $fail_fastq \\
            $args \\
            2> >(tee ${prefix}.fastp.log >&2)

        cat <<-END_VERSIONS > versions.yml
        "${task.process}":
            fastp: \$(fastp --version 2>&1 | sed -e "s/fastp //g")
        END_VERSIONS
        """
    } else {
        def merge_fastq = save_merged ? "-m --merged_out ${prefix}.merged.fastq.gz" : ''
        """
        [ ! -f  ${prefix}_1.fastq.gz ] && ln -sf ${reads[0]} ${prefix}_1.fastq.gz
        [ ! -f  ${prefix}_2.fastq.gz ] && ln -sf ${reads[1]} ${prefix}_2.fastq.gz
        fastp \\
            --in1 ${prefix}_1.fastq.gz \\
            --in2 ${prefix}_2.fastq.gz \\
            --out1 ${prefix}_1.fastp.fastq.gz \\
            --out2 ${prefix}_2.fastp.fastq.gz \\
            --json ${prefix}.fastp.json \\
            --html ${prefix}.fastp.html \\
            $adapter_list \\
            $fail_fastq \\
            $merge_fastq \\
            --thread $task.cpus \\
            --detect_adapter_for_pe \\
            $args \\
            2> >(tee ${prefix}.fastp.log >&2)

        cat <<-END_VERSIONS > versions.yml
        "${task.process}":
            fastp: \$(fastp --version 2>&1 | sed -e "s/fastp //g")
        END_VERSIONS
        """
    }

    stub:
    def prefix              = task.ext.prefix ?: "${meta.id}"
    def is_single_output    = task.ext.args?.contains('--interleaved_in') || meta.single_end
    def touch_reads         = is_single_output ? "${prefix}.fastp.fastq.gz" : "${prefix}_1.fastp.fastq.gz ${prefix}_2.fastp.fastq.gz"
    def touch_merged        = (!is_single_output && save_merged) ? "touch ${prefix}.merged.fastq.gz" : ""
    """
    touch $touch_reads
    touch "${prefix}.fastp.json"
    touch "${prefix}.fastp.html"
    touch "${prefix}.fastp.log"
    $touch_merged

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        fastp: \$(fastp --version 2>&1 | sed -e "s/fastp //g")
    END_VERSIONS
    """
}
