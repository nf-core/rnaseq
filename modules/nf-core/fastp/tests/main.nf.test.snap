{
    "fastp test_fastp_interleaved_json": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "test.fastp.json:md5,168f516f7bd4b7b6c32da7cba87299a4"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-01-17T18:08:06.123035"
    },
    "test_fastp_paired_end_merged-for_stub_match": {
        "content": [
            [
                [
                    "test_1.fastp.fastq.gz",
                    "test_2.fastp.fastq.gz"
                ],
                "test.fastp.html",
                "test.fastp.json",
                "test.fastp.log",
                "test.merged.fastq.gz",
                "{id=test, single_end=false}"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-01-17T18:10:13.467574"
    },
    "versions_interleaved": {
        "content": [
            [
                "versions.yml:md5,48ffc994212fb1fc9f83a74fa69c9f02"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-02-01T11:56:24.615634793"
    },
    "test_fastp_single_end_json": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "test.fastp.json:md5,c852d7a6dba5819e4ac8d9673bedcacc"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-01-17T18:06:00.223817"
    },
    "versions_paired_end": {
        "content": [
            [
                "versions.yml:md5,48ffc994212fb1fc9f83a74fa69c9f02"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-02-01T11:55:42.333545689"
    },
    "test_fastp_paired_end_match": {
        "content": [
            [
                [
                    "test_1.fastp.fastq.gz",
                    "test_2.fastp.fastq.gz"
                ],
                "test.fastp.html",
                "test.fastp.json",
                "test.fastp.log",
                "{id=test, single_end=false}"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-02-01T12:03:06.431833729"
    },
    "test_fastp_interleaved-_match": {
        "content": [
            [
                "test.fastp.fastq.gz",
                "test.fastp.html",
                "test.fastp.json",
                "test.fastp.log",
                "{id=test, single_end=true}"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-02-01T12:03:37.827323085"
    },
    "test_fastp_paired_end_merged_match": {
        "content": [
            [
                [
                    "test_1.fastp.fastq.gz",
                    "test_2.fastp.fastq.gz"
                ],
                "test.fastp.html",
                "test.fastp.json",
                "test.fastp.log",
                "test.merged.fastq.gz",
                "{id=test, single_end=false}"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-02-01T12:08:44.496251446"
    },
    "versions_single_end_stub": {
        "content": [
            [
                "versions.yml:md5,48ffc994212fb1fc9f83a74fa69c9f02"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-02-01T11:55:27.354051299"
    },
    "versions_interleaved-stub": {
        "content": [
            [
                "versions.yml:md5,48ffc994212fb1fc9f83a74fa69c9f02"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-02-01T11:56:46.535528418"
    },
    "versions_single_end_trim_fail": {
        "content": [
            [
                "versions.yml:md5,48ffc994212fb1fc9f83a74fa69c9f02"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-02-01T11:59:03.724591407"
    },
    "test_fastp_paired_end-for_stub_match": {
        "content": [
            [
                [
                    "test_1.fastp.fastq.gz",
                    "test_2.fastp.fastq.gz"
                ],
                "test.fastp.html",
                "test.fastp.json",
                "test.fastp.log",
                "{id=test, single_end=false}"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-01-17T18:07:15.398827"
    },
    "versions_paired_end-stub": {
        "content": [
            [
                "versions.yml:md5,48ffc994212fb1fc9f83a74fa69c9f02"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-02-01T11:56:06.50017282"
    },
    "versions_single_end": {
        "content": [
            [
                "versions.yml:md5,48ffc994212fb1fc9f83a74fa69c9f02"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-02-01T11:55:07.67921647"
    },
    "versions_paired_end_merged_stub": {
        "content": [
            [
                "versions.yml:md5,48ffc994212fb1fc9f83a74fa69c9f02"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-02-01T11:59:47.350653154"
    },
    "test_fastp_interleaved-for_stub_match": {
        "content": [
            [
                "test.fastp.fastq.gz",
                "test.fastp.html",
                "test.fastp.json",
                "test.fastp.log",
                "{id=test, single_end=true}"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-01-17T18:08:06.127974"
    },
    "versions_paired_end_trim_fail": {
        "content": [
            [
                "versions.yml:md5,48ffc994212fb1fc9f83a74fa69c9f02"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-02-01T11:59:18.140484878"
    },
    "test_fastp_single_end-for_stub_match": {
        "content": [
            [
                "test.fastp.fastq.gz",
                "test.fastp.html",
                "test.fastp.json",
                "test.fastp.log",
                "{id=test, single_end=true}"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-01-17T18:06:00.244202"
    },
    "test_fastp_single_end-_match": {
        "content": [
            [
                "test.fastp.fastq.gz",
                "test.fastp.html",
                "test.fastp.json",
                "test.fastp.log",
                "{id=test, single_end=true}"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-02-01T11:57:30.791982648"
    },
    "versions_paired_end_merged_adapterlist": {
        "content": [
            [
                "versions.yml:md5,48ffc994212fb1fc9f83a74fa69c9f02"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-02-01T12:05:37.845370554"
    },
    "versions_paired_end_merged": {
        "content": [
            [
                "versions.yml:md5,48ffc994212fb1fc9f83a74fa69c9f02"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-02-01T11:59:32.860543858"
    },
    "test_fastp_single_end_trim_fail_json": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "test.fastp.json:md5,9a7ee180f000e8d00c7fb67f06293eb5"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-01-17T18:08:41.942317"
    }
}